LocalProviderPath failed to find a provider binary

### What happened?

Have some evidence in https://github.com/pulumi/pulumi-aws/issues/3700 that LocalProviderPath proceeds with the test when the provider is not found at a given path. Would prefer this to fail right away since the subsequent test failure is more cryptic, or sometimes the test, especially YAML test, may run against a different version of the provider which makes it give a false sense of security for testing the provider-under-test.

```
	test := pulumitest.NewPulumiTest(t, "no-code-lambda",
		opttest.LocalProviderPath("aws", filepath.Join(getCwd(t), "..", "bin")),
	)
```

### Example

See above.

### Output of `pulumi about`

N/A

### Additional context

N/A

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

Sure it's still a problem.

Take https://github.com/pulumi/pulumi-terraform-module-provider/blob/6478a47b41507867f9dfbea7f26c96c4bf75d83c/tests/acc_test.go#L46

If you inject fault it will fail with:

```
=== RUN   Test_RandMod_TypeScript/pulumi_package_add
    acc_test.go:1369: Failed to run pulumi package add
        Exit code: 255
        Error: exit status 255
        
        error: failed to get schema: could not find file /Users/anton/code/pulumi-terraform-module/bin/pulumi-resource-terraform-module
    acc_test.go:1372: 
        	Error Trace:	/Users/anton/code/pulumi-terraform-module/tests/acc_test.go:1372
        	            				/Users/anton/code/pulumi-terraform-module/tests/acc_test.go:63
        	Error:      	Received unexpected error:
        	            	exit status 255
        	Test:       	Test_RandMod_TypeScript/pulumi_package_add

error: Running program '/private/var/folders/gd/3ncjb1lj5ljgk8xl5ssn_gvc0000gn/T/com.apple.shortcuts.mac-helper/Test_RandMod_TypeScript4215673288/001/randmod-program/index.ts' failed with an unhandled exception:
TSError: ⨯ Unable to compile TypeScript:
index.ts(2,26): error TS2307: Cannot find module '@pulumi/randmod' or its corresponding type declarations.

Resources:
    + 1 to create
    acc_test.go:67: failed to preview update: failed to run preview: exit status 255
        code: 255
        stdout: Previewing update (test):
        + pulumi:pulumi:Stack: (create)
            [urn=urn:pulumi:test::ts-randmod-program::pulumi:pulumi:Stack::ts-randmod-program-test]
        Resources:
            + 1 to create

```

There's an opportunity to fail faster and say LocalProviderPath is pointing to a non-existing path.

--- a/pulumitest/newStack.go
+++ b/pulumitest/newStack.go
@@ -107,6 +107,11 @@ func (pt *PulumiTest) NewStack(t PT, stackName string, opts ...optnewstack.NewSt
 			if err != nil {
 				ptFatalF(t, "failed to get absolute path for %s: %s", relPath, err)
 			}
+			_, err = os.Stat(absPath)
+			if err != nil {
+				ptFatalF(t, "failed to find binary for provider %q: %s", name, err)
+				return nil
+			}
 
 			found := false
 			for idx := range providerPlugins {

--- a/pulumitest/newStack_test.go
+++ b/pulumitest/newStack_test.go
@@ -1,0 +1,21 @@
+package pulumitest_test
+
+import (
+	"path/filepath"
+	"testing"
+
+	"github.com/pulumi/providertest/pulumitest"
+	"github.com/pulumi/providertest/pulumitest/opttest"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMissingProviderBinaryPath(t *testing.T) {
+	t.Parallel()
+
+	tt := mockT{T: t}
+	pulumitest.NewPulumiTest(&tt, filepath.Join("testdata", "yaml_program"),
+		opttest.LocalProviderPath("gcp", filepath.Join("provider_directory", "file_that_does_not_exist")),
+	)
+
+	assert.True(t, tt.Failed(), "expected test to fail")
+}

